pkg/streamingpromql/functions/common.go

```diff
@@ -22,6 +24,11 @@ func DropSeriesName(seriesMetadata []types.SeriesMetadata, _ *limiting.MemoryCon
 		seriesMetadata[i].Labels = seriesMetadata[i].Labels.DropMetricName()
 	}
 
+	if hasDuplicate, example := types.HasDuplicateSeries(seriesMetadata); hasDuplicate {
```

Comment:
Should this check not be done in query.go when we get all of the series names?
It would help catch other operators incorrectly handling series labels such as aggregations. Although that's more likely a bug and unnecessary overhead. Or perhaps we do it in query.go in `convertFunctionCallToOperator` so that we apply it to all functions that can manipulate labels. Otherwise if we manipulate labels differently in a future function we may forget to apply this check.

Reply:
>Should this check not be done in query.go when we get all of the series names? It would help catch other operators incorrectly handling series labels such as aggregations. Although that's more likely a bug and unnecessary overhead.

This wouldn't be a thorough-enough check - duplicate series are not permitted in any intermediate stage, not just the end result. And yep, I think it would be unnecessary overhead and a bug.

> Or perhaps we do it in query.go in `convertFunctionCallToOperator` so that we apply it to all functions that can manipulate labels. Otherwise if we manipulate labels differently in a future function we may forget to apply this check.

I'm thinking something more along these lines makes sense, especially given [this realisation](https://github.com/grafana/mimir/pull/9019#issuecomment-2295413980) over the weekend. 

Given we need similar checking and merging logic in many places[^1], I'm wondering if this should be done in an operator that wraps operators that manipulate labels. This wrapping operator would be responsible for:
- merging the results from different input series that map to the same output series
- rejecting queries where results from different input series conflict at a single timestamp

We can then apply this operator whenever the expression manipulates labels.

The other option would be to add this logic to `FunctionOverInstantVector` and `FunctionOverRangeVector`, and possibly other places in the future, but I think that will make them more complex to reason about and unnecessarily duplicate some logic that can live in one place.

What do you think?

[^1]: all the functions over instant vectors and range vectors, `label_replace`, `label_join`, unary negation, possibly others?

Reply:
I think a wrapper operator makes the most sense. I think the only reason not to would be if we have an opportunity in a downstream operator to detect such a conflict. However, we can likely do both and if we have an opportunity for such an optimisation we can percolate the error back up.